### ISSUE TITLE
Efficiency: Better Way to Calculate Entropy?

### ISSUE THREAD

**Issue by lafemmephile** [lafemmephile/core#34]
# Purpose
To find a more efficient (if it exists) way to calculate the entropy of a *video file*.

**Comment by lafemmephile** [comment]
# Performance Tests
As can be seen in the results section below, it seems as though `moviepy` is faster at *iterating frames* than `pyav`, or at least my implementation is faster with `moviepy` ...

# Code
The code used for the performance tests.
```python
import pathlib
from femmephile.video.entropy import video_entropy
from femmephile.video.utils import gen_video_paths
import av

DATA_PATH = pathlib.Path("/usr/local/src/core/notebooks/data/jan_2023/")

for path in gen_video_paths(DATA_PATH):
    get_ipython().run_line_magic('time', 'entropy = video_entropy(str(path))')
    print(f"{entropy}Sh {path}")

def av_entropy(video_path):
    with av.open(video_path) as container:
        total_frames = 0
        total_entropy = 0
        stream = container.streams.video[0]
        for frame_indx, frame in enumerate(container.decode(stream)):
            total_entropy += frame.to_image().entropy()
            total_frames += 1
    return total_entropy / total_frames, total_frames
       
for path in gen_video_paths(DATA_PATH):
    get_ipython().run_line_magic('time', 'entropy = av_entropy(str(path))')
    print(f"{entropy}Sh {path}")
```

# Results
+ `moviepy`
```console
CPU times: user 2.82 s, sys: 867 ms, total: 3.69 s
Wall time: 3.8 s
(8.68619022859487, 912)Sh /usr/local/src/core/notebooks/data/jan_2023/1675118416363596.webm
CPU times: user 128 ms, sys: 23.9 ms, total: 152 ms
Wall time: 194 ms
(8.955642209347312, 77)Sh /usr/local/src/core/notebooks/data/jan_2023/1672229982266113.webm
CPU times: user 1.15 s, sys: 358 ms, total: 1.51 s
Wall time: 1.56 s
(7.844446142104237, 750)Sh /usr/local/src/core/notebooks/data/jan_2023/1675189732439918.webm
CPU times: user 749 ms, sys: 164 ms, total: 913 ms
Wall time: 1.02 s
(8.954791372394181, 294)Sh /usr/local/src/core/notebooks/data/jan_2023/1675126240383278.webm
CPU times: user 3.34 s, sys: 720 ms, total: 4.06 s
Wall time: 4.17 s
(9.200359094701582, 1383)Sh /usr/local/src/core/notebooks/data/jan_2023/1675125561832690.webm
CPU times: user 806 ms, sys: 347 ms, total: 1.15 s
Wall time: 1.24 s
(9.05173878423511, 286)Sh /usr/local/src/core/notebooks/data/jan_2023/1675120041290161.webm
CPU times: user 886 ms, sys: 68.1 ms, total: 954 ms
Wall time: 1.05 s
(9.071143639775952, 235)Sh /usr/local/src/core/notebooks/data/jan_2023/1675158373239750.webm
CPU times: user 1.54 s, sys: 315 ms, total: 1.86 s
Wall time: 1.94 s
(9.279875723700625, 624)Sh /usr/local/src/core/notebooks/data/jan_2023/1675117108875152.webm
CPU times: user 191 ms, sys: 45.5 ms, total: 237 ms
Wall time: 290 ms
(8.17564296799289, 138)Sh /usr/local/src/core/notebooks/data/jan_2023/1675125674121660.webm
```

+ `pyav`
```console
CPU times: user 10.2 s, sys: 3.21 s, total: 13.4 s
Wall time: 10.7 s
(8.686093776017062, 911)Sh /usr/local/src/core/notebooks/data/jan_2023/1675118416363596.webm
CPU times: user 249 ms, sys: 4.31 ms, total: 254 ms
Wall time: 253 ms
(8.954379860124266, 75)Sh /usr/local/src/core/notebooks/data/jan_2023/1672229982266113.webm
CPU times: user 3.74 s, sys: 76.9 ms, total: 3.81 s
Wall time: 2.81 s
(7.8445833716139175, 749)Sh /usr/local/src/core/notebooks/data/jan_2023/1675189732439918.webm
CPU times: user 3.76 s, sys: 772 ms, total: 4.53 s
Wall time: 3.2 s
(8.936081699278168, 293)Sh /usr/local/src/core/notebooks/data/jan_2023/1675126240383278.webm
CPU times: user 9.36 s, sys: 663 ms, total: 10 s
Wall time: 7.58 s
(9.200359094701582, 1383)Sh /usr/local/src/core/notebooks/data/jan_2023/1675125561832690.webm
CPU times: user 3.16 s, sys: 913 ms, total: 4.07 s
Wall time: 3.28 s
(9.05124743159674, 285)Sh /usr/local/src/core/notebooks/data/jan_2023/1675120041290161.webm
CPU times: user 2.59 s, sys: 752 ms, total: 3.35 s
Wall time: 2.68 s
(9.072877638353399, 234)Sh /usr/local/src/core/notebooks/data/jan_2023/1675158373239750.webm
CPU times: user 5.57 s, sys: 1.57 s, total: 7.13 s
Wall time: 5.48 s
(9.279875723700625, 624)Sh /usr/local/src/core/notebooks/data/jan_2023/1675117108875152.webm
CPU times: user 704 ms, sys: 5.15 ms, total: 709 ms
Wall time: 504 ms
(8.176434717859115, 137)Sh /usr/local/src/core/notebooks/data/jan_2023/1675125674121660.webm
```

**Comment by smttsp** [comment]
Regardless of the algorithm, neither of them is acceptable speed, to be honest. 

How long are the videos? 

If you update your code and make it something like the following:

```
import cv2

# Open the video file
video_path = 'path/to/your/video/file.mp4'
cap = cv2.VideoCapture(video_path)

# Check if the video file is successfully opened
if not cap.isOpened():
    print("Error: Could not open video file.")
    exit()

total_frames = 0
total_entropy = 0

while True:
    # Read the next frame
    ret, frame = cap.read()

    # Break the loop if the video is over
    if not ret:
        break

    # Get the type of the frame (0 for I-frame, 1 for P-frame, 2 for B-frame)
    frame_type = int(cap.get(cv2.CAP_PROP_POS_FRAMETYPE))

    # Filter out I-frames
    if frame_type == 0:
        # Process and display the I-frame
        total_entropy += frame.to_image().entropy()
        total_frames += 1

# Release the video capture object and close all windows
cap.release()
cv2.destroyAllWindows()
```

where you only keep track of i-frames, how long will that be?

---
You can do the same thing for the other algorithms. For example, this should work for your code

```
def av_entropy(video_path):
    with av.open(video_path) as container:
        total_frames = 0
        total_entropy = 0
        stream = container.streams.video[0]
        for frame_indx, frame in enumerate(container.decode(stream)):
            # Check if the frame is an I-frame (intra-coded frame)
            if frame.pict_type == 'I':
                total_entropy += frame.to_image().entropy()
                total_frames += 1
    return total_entropy / total_frames, total_frames
```
---

If you focus on only i-frames, your processing will reduce by 25-30x and I don't think entropy will change significantly if you ignore p- and b-frames.

**Comment by smttsp** [comment]
```
CPU times: user 2.82 s, sys: 867 ms, total: 3.69 s
Wall time: 3.8 s
(8.68619022859487, 912)Sh /usr/local/src/core/notebooks/data/jan_2023/1675118416363596.webm
```

for 912 frames which should be about 30-35 sec video, it takes about 4 seconds. So, to process 1 hour video, this algorithm would take 6 mins or so which is not acceptable in any scenario. Besides, this is only one of the calculations, there will be a bunch of other calculations. 

Hence, I would suggest that we should sample, such as sampling up to `MAX_NUMBER=100` frames per video.

**Comment by smttsp** [comment]
@lafemmephile, I tried your code with a single video, so no big claims in terms of entropy, but I am quite confident that it all frames should be pretty much equal to only I-frames. 

Below are my results:
I kept track of time as well as did a memory profiling using scalene
`scalene` is a (really nice idea but not great code) memory and cpu time profiler. 

```
def av_entropy(video_path):
    with av.open(video_path) as container:
        total_frames = 0
        total_entropy = 0
        stream = container.streams.video[0]
        for frame_indx, frame in enumerate(container.decode(stream)):
            total_entropy += frame.to_image().entropy()
            total_frames += 1
    return total_entropy / total_frames, total_frames
```
time 13.292
(8.866303728551637, 1358)

<img width="884" alt="Screenshot 2023-10-05 at 9 50 07 PM" src="https://github.com/lafemmephile/core/assets/4594945/5a8888e0-9b93-41ac-8bb7-28c8e7a950dc">

(`total_entropy += frame.to_image().entropy()` uses 3x time compared to `for frame_indx, frame in enumerate(container.decode(stream)))

---

I-frames only
```
def av_entropy(video_path):
    t1 = time.time()

    with av.open(video_path) as container:
        total_frames = 0
        total_entropy = 0
        stream = container.streams.video[0]
        for frame_indx, frame in enumerate(container.decode(stream)):
            # Check if the frame is an I-frame (intra-coded frame)
            if frame.pict_type == 'I':
                total_entropy += frame.to_image().entropy()
                total_frames += 1
    print("time", round(time.time() - t1, 3))
    return total_entropy / total_frames, total_frames
```
time 2.789
(8.863757293458471, 8)
 
<img width="1253" alt="Screenshot 2023-10-05 at 9 48 07 PM" src="https://github.com/lafemmephile/core/assets/4594945/b29348cc-7ffc-4008-a169-3615184bee3c">

---

I tried the random access of opencv which is definitely not good. So, regardless, we need to iterate through the entire video

**Comment by smttsp** [comment]
opencv version of reading all frames and getting entropy:

```
def cv2_entropy(video_path):
    t1 = time.time()

    cap = cv2.VideoCapture(video_path)

    if not cap.isOpened():
        print("Error: Could not open video file.")
        exit()

    entropies = []
    while True:
        ret, frame = cap.read()
        if not ret:
            break
        entropies.append(Image.fromarray(frame).entropy())

    cap.release()

    print("time", round(time.time() - t1, 3))

    return np.mean(entropies), len(entropies)
```

> time 4.879
> (8.8663, 1358)

<img width="1113" alt="Screenshot 2023-10-05 at 10 38 55 PM" src="https://github.com/lafemmephile/core/assets/4594945/20a9c272-c976-48fa-8b59-0855585f44a4">

--- 
Reading one frame per second. I don't think we can use I-frames in opencv. 

```
    entropies = []
    idx = 0
    while True:
        ret, frame = cap.read()
        if not ret:
            break

        if idx % frame_rate == 0:
            entropies.append(Image.fromarray(frame).entropy())
        idx += 1
    print("time", round(time.time() - t1, 3))
    return np.mean(entropies), len(entropies)
```

> time 1.57
> (8.86526174671925, 24)

and `scalene` visualization:

<img width="1172" alt="Screenshot 2023-10-05 at 10 43 20 PM" src="https://github.com/lafemmephile/core/assets/4594945/1e560b25-3526-4726-9e7f-1146a6c0c5e7">

---
So, as seen, we now ended up with only video reading time which is a majority of the time. 

---
The last thing we can do is limit the number of frames to something like 100, so that worst case scenario, we would spend a few seconds per video (which is still quite high, we need multi-threading and a bunch of other stuff).

**Comment by smttsp** [comment]
> why an intelligent frame sampling strategy is going to be a necessary aspect of any analysis of video data (e.g. measuring entropy).

more importantly, I think we need to make sure that we iterate through the video frames only once as it is a quite expensive operation. 

it should be like this:

```
for idx, frame in enumerate(video):  <-- super slow
     calculate_entropy()
     calculate_other_thing1()
     calculate_other_thing2() 
     
     store_all_the_findings(idx, findings)
```

in other words, suppose that we are looking for 5 things in a video, the `super slow` step should be done only once. Otherwise, for each video, we will spend 5x time on that step. 

---
To be honest, I still don't see how 1000+ videos can be handled. Maybe 

- a really nice caching mechanism (we could store information about videos in a cache file, so in the second run, things would load much faster),
- multiprocessing, 
- beautiful visualizations, i.e. informing the user of the video processing progress in a separate tab (one tab for `ImageLab` and another one for `VideoLab`)

**Comment by lafemmephile** [comment]
> in other words, suppose that we are looking for 5 things in a video, the super slow step should be done only once. Otherwise, for each video, we will spend 5x time on that step.

What about the *caching* aspect of the [VideoSampler](https://github.com/LemurPwned/video-sampler) code? It just pulls out the desired frames, writes them to disk in some directory (labeled based on the *parent video file*), and then these frames get reused over and over?

But yes I do see your meaning ... the *COST* of the *frame iteration* is **non-trivial** ... and thus any code that seeks to analyze frames (e.g. `Videolab`) must implement an approach that seeks to minimize this cost as much as possible.

**Comment by smttsp** [comment]
>  It just pulls out the desired frames, writes them to disk in some directory (labeled based on the parent video file), and then these frames get reused over and over

I think reading frames from disk may be slower than decoding.

**Comment by lafemmephile** [comment]
@smttsp Maybe what we can do is create a repository just for testing various *frame sampling* strategies ... where we develop the different approaches, find appropriate test data, and then write the tests to gather the *experimental* results of how different *frame sampling* strategies perform?

**Comment by smttsp** [comment]
makes sense. I have copied @LemurPwned's repo and built some stuff on it. We can keep working on that repo if you want: https://github.com/smttsp/cleanvision/tree/videolab

Or we can work on the original repo or your repo. I am fine with any of them

**Comment by LemurPwned** [comment]
Hey guys I see I've been called out :D 
One thing -- setting the the nokey decoder before the decoding stream starts seems to be faster than checking for I frames during the decoding -- see [here](https://github.com/LemurPwned/video-sampler/blob/02eff3e07edb6e3b6c05c71f7ca7ab735563dea5/video_sampler/sampler.py#L28). And decoding all frames vs only keyframes is night and day difference. Any seek ops are also going to be slow.

Another issue is that to_image is rather slow (it's does the actual transform under the hood). I haven't tested it but perhaps `to_ndarray` is much faster, but since image hash didn't support it, I didn't bother. 

@smttsp -- those scalene visualisations look great!

**Comment by lafemmephile** [comment]
> Hey guys I see I've been called out :D One thing -- setting the the nokey decoder before the decoding stream starts seems to be faster than checking for I frames during the decoding -- see [here](https://github.com/LemurPwned/video-sampler/blob/02eff3e07edb6e3b6c05c71f7ca7ab735563dea5/video_sampler/sampler.py#L28). And decoding all frames vs only keyframes is night and day difference. Any seek ops are also going to be slow.
> 
> Another issue is that `to_image` is rather slow (it's does the actual transform under the hood). I haven't tested it but perhaps `to_ndarray` is much faster, but since image hash didn't support it, I didn't bother.
> 
> @smttsp -- those scalene visualisations look great!

@smttsp @LemurPwned Awesome, these are the kinds of *performant choices* I want to explore in https://github.com/lafemmephile/frame_sampling. There we can systematically implement, test, and experiment with different frame sampling implementations/strategies.
